File: widget/lib_network/src/main/java/com/yyxnb/network/BaseViewModel.java
package com.yyxnb.network;

import android.arch.lifecycle.DefaultLifecycleObserver;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.content.Context;

import com.yyxnb.common.action.CommonAction;
import com.yyxnb.widget.WidgetManager;
import com.yyxnb.widget.interfaces.IData;
import com.yyxnb.network.rx.BaseHttpSubscriber;
import com.yyxnb.network.rx.RetryWithDelay;
import com.yyxnb.rxtool.DisposablePool;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

@SuppressWarnings("rawtypes")
public abstract class BaseViewModel extends ViewModel implements DefaultLifecycleObserver, CommonAction {

    public final MutableLiveData<Status> status = new MutableLiveData<>();
    private final List<Integer> mDisposable = new ArrayList<>();

    @Override
    public Context getContext() {
        return WidgetManager.getInstance().getContext();
    }

    public <T extends IData> BaseHttpSubscriber<T> request(Flowable<T> flowable) {
        final BaseHttpSubscriber<T> baseHttpSubscriber = new BaseHttpSubscriber<T>();
        //RxJava Subscriber回调
        flowable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(baseHttpSubscriber);
        return baseHttpSubscriber;
    }

    public <T extends IData> LiveData<T> request(Observable<T> observable) {
        return new ObservableLiveData<>(observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).retryWhen(new RetryWithDelay()));
    }

    public <T extends IData> void launchOnlyResult(
            Observable<T> call,
            OnHandleException<T> onHandleException
    ) {
        status.postValue(Status.LOADING);
        call.subscribeOn(Schedulers.io())
                .retryWhen(new RetryWithDelay())
                .subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<T>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        mDisposable.add(call.hashCode());
                        DisposablePool.get().add(d, call.hashCode());
                    }

                    @Override
                    public void onNext(T t) {
                        if (t.isSuccess()) {
                            status.postValue(Status.SUCCESS);
                            onHandleException.success(t);
                        } else {
                            status.postValue(Status.ERROR);
                            onHandleException.error(t.getCode());
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        status.postValue(Status.ERROR);
                        onHandleException.error(e.getMessage());
                        e.printStackTrace();
                    }

                    @Override
                    public void onComplete() {
                        status.postValue(Status.COMPLETE);
                    }
                });
    }

    public interface OnHandleException<T> {

        void success(T data);

        void error(String msg);

    }

    @Override
    protected void onCleared() {
        super.onCleared();
        for (Integer i : mDisposable) {
            DisposablePool.get().remove(i);
        }
        mDisposable.clear();
    }
}

File: lib_network/src/main/java/com/yyxnb/network/BaseViewModel.kt
package com.yyxnb.network

import android.view.View
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.yyxnb.common.interfaces.IData
import com.yyxnb.common.utils.log.LogUtils
import com.yyxnb.http.exception.ResponseThrowable
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

abstract class BaseViewModel : ViewModel() {

    val status = MutableLiveData<Status>()

    //重试的监听
    var listener: View.OnClickListener? = null

    val mScope: CoroutineScope by lazy {
        CoroutineScope(SupervisorJob() + Dispatchers.Main)
    }

    /**
     * 所有网络请求都在 viewModelScope 域中启动，当页面销毁时会自动
     * 调用ViewModel的  #onCleared 方法取消所有协程
     */
    fun launchUI(block: suspend CoroutineScope.() -> Unit) = mScope.launch { block() }

    /**
     * 用流的方式进行网络请求
     */
    fun <T> launchFlow(block: suspend () -> T): Flow<T> {
        return flow {
            emit(block())
        }
    }

    /**
     * 过滤请求结果，其他全抛异常
     * @param block 请求体
     * @param success 成功回调
     * @param error 失败回调
     * @param complete  完成回调（无论成功失败都会调用）
     * @param view
     *
     **/
    fun <T> launchOnlyResult(
            block: suspend CoroutineScope.() -> IData<T>,
            //成功
            success: (T) -> Unit = {},
            //错误 根据错误进行不同分类
            error: (Throwable) -> Unit = {
                //UnknownHostException 1：服务器地址错误；2：网络未连接
                reTry()
            },
            //完成
            complete: () -> Unit = {},
            //重试
            reTry: () -> Unit = {}
    ) {
        status.postValue(Status.LOADING)
        //正式请求接口
        launchUI {
            //异常处理
            handleException(
                    //调用接口
                    { withContext(Dispatchers.IO) { block() } },
                    { res ->
                        //接口成功返回
                        executeResponse(res) {
                            success(it)
                        }
                    },
                    {
                        status.postValue(Status.ERROR)
                        //接口失败返回
                        error(it)
                    },
                    {
                        status.postValue(Status.COMPLETE)
                        //接口完成
                        complete()
                    }
            )
        }
    }


    /**
     * 请求结果过滤
     */
    private suspend fun <T> executeResponse(
            response: IData<T>,
            success: suspend CoroutineScope.(T) -> Unit
    ) {
        coroutineScope {
            //接口成功返回后判断是否是增删改查成功，不满足的话，返回异常
            if (response.isSuccess) {
                status.postValue(Status.SUCCESS)
                success(response.result!!)
            } else {
                status.postValue(Status.ERROR)
                //状态码错误
                throw ResponseThrowable(
                        response.code!!,
                        response.msg!!
                )
            }
        }
    }

    /**
     * 异常统一处理
     */
    private suspend fun <T> handleException(
            block: suspend CoroutineScope.() -> IData<T>,
            success: suspend CoroutineScope.(IData<T>) -> Unit,
            error: suspend CoroutineScope.(Throwable) -> Unit,
            complete: suspend CoroutineScope.() -> Unit
    ) {
        coroutineScope {
            try {
                success(block())
            } catch (e: Throwable) {
                error(e)
                e.printStackTrace()
            } finally {
                complete()
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        mScope.cancel()
        listener = null
    }
}